network-store-iidm-impl/src/main/java/com/powsybl/network/store/iidm/impl/NodeBreakerViewImpl.java

```diff
@@ -125,43 +126,108 @@ public void traverse(int[] nodes, VoltageLevel.NodeBreakerView.TopologyTraverser
-                result = traverser.traverse(node, null, nextNode);
-            } else {
-                throw new AssertionError();
+    private boolean traverseFromNode(Graph<Integer, Edge> graph, int node, TraversalType traversalType,
+                                     VoltageLevel.NodeBreakerView.TopologyTraverser traverser, Set<Integer> done) {
+        if (traversalType == TraversalType.DEPTH_FIRST) {   // traversal by depth first
+            if (done.contains(node)) {
+                return true;
             }
-            if (result == TraverseResult.CONTINUE) {
-                if (!traverseFromNode(graph, nextNode, traverser, done)) {
+            done.add(node);
+
+            for (Edge edge : graph.edgesOf(node)) {
+                NodeBreakerBiConnectable biConnectable = edge.getBiConnectable();
+                int nextNode = biConnectable.getNode1() == node ? biConnectable.getNode2() : biConnectable.getNode1();
+                TraverseResult result;
+                if (done.contains(nextNode)) {
+                    continue;
+                }
+                if (biConnectable instanceof SwitchAttributes) {
+                    result = traverseSwitch(traverser, biConnectable, node, nextNode);
+                } else if (biConnectable instanceof InternalConnectionAttributes) {
+                    result = traverser.traverse(node, null, nextNode);
+                } else {
+                    throw new AssertionError();
+                }
+                if (result == TraverseResult.CONTINUE) {
+                    if (!traverseFromNode(graph, nextNode, traversalType, traverser, done)) {
+                        return false;
+                    }
+                } else if (result == TraverseResult.TERMINATE_TRAVERSER) {
                     return false;
                 }
-            } else if (result == TraverseResult.TERMINATE_TRAVERSER) {
-                return false;
             }
+        } else {   // traversal by breadth first
+            boolean keepGoing = true;
+            Set<Edge> encounteredEdges = new HashSet<>();
+
+            LinkedList<Integer> vertexToTraverse = new LinkedList<>();
+            vertexToTraverse.offer(node);
+            while (!vertexToTraverse.isEmpty()) {
+                int firstV = vertexToTraverse.getFirst();
+                vertexToTraverse.poll();
+                if (done.contains(firstV)) {
+                    continue;
+                }
+                done.add(firstV);
+
+                Set<Edge> adjacentEdges = graph.edgesOf(firstV);
+
+                for (Edge edge : adjacentEdges) {
+                    if (encounteredEdges.contains(edge)) {
+                        continue;
+                    }
+                    encounteredEdges.add(edge);
+
+                    NodeBreakerBiConnectable biConnectable = edge.getBiConnectable();
+                    int node1 = biConnectable.getNode1();
+                    int node2 = biConnectable.getNode2();
+
```

Comment:
we can also use int nextNode similarly to the DFS to avoid the following if clause